feat: implement and calculate the current key period using pallas

falcucci opened the pull request:
## Pre-submit checklist

- Branch
  - [x] Tests are provided (if possible)
  - [x] Crates versions are updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] No clippy warnings in the CI
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested
- Documentation
  - [ ] Update README file (if relevant)
  - [ ] Update documentation website (if relevant)
  - [ ] Add dev blog post (if relevant)

falcucci committed:
feat: add kes period

- update the source of dependencies `pallas-addresses`, `pallas-codec`, `pallas-network`, `pallas-primitives`, `pallas-traverse` to a git repository with branch `feat/genesis-config`
- add a new method `calculate_kes_period` to fetch chain point and genesis config through local statequery
- implement logic in the `calculate_kes_period` method to calculate the kes period based on retrieved data

falcucci committed:
refactor: remove fallback field usage

- remove the `fallback` field from `pallaschainobserver`
- update the `pallaschainobserver::new` method signature to no longer accept `fallback` argument in `pallas_observer.rs`
- update the test functions in the `mod tests` section to no longer use `fallback` in `pallas_observer.rs`
- remove the `fallback` field and related code in `mithrilinfrastructure` in `infrastructure.rs`

falcucci committed:
refactor: streamline code formatting in various modules

- remove unnecessary imports in `chain_observer/pallas_observer.rs`
- fix formatting issues in the code
- simplify the `get_fake_utxo_by_address` function definition

scarmuega commented on mithril-common/Cargo.toml:
```diff
@@ -35,11 +35,16 @@ kes-summed-ed25519 = { version = "0.2.1", features = [
-pallas-network = { version = "0.23.0", optional = true }
-pallas-primitives = { version = "0.23.0", optional = true }
-pallas-traverse = { version = "0.23.0", optional = true }
+# pallas-addresses = { version = "0.23.0", optional = true }
+pallas-addresses = { git = "https://github.com/falcucci/pallas.git", branch = "feat/genesis-config", optional = true }
+# pallas-codec = { version = "0.23.0", optional = true }
+pallas-codec = { git = "https://github.com/falcucci/pallas.git", branch = "feat/genesis-config", optional = true }
+# pallas-network = { version = "0.23.0", optional = true }
+pallas-network = { git = "https://github.com/falcucci/pallas.git", branch = "feat/genesis-config", optional = true }
+# pallas-primitives = { version = "0.23.0", optional = true }
+pallas-primitives = { git = "https://github.com/falcucci/pallas.git", branch = "feat/genesis-config", optional = true }
+# pallas-traverse = { version = "0.23.0", optional = true }
+pallas-traverse = { git = "https://github.com/falcucci/pallas.git", branch = "feat/genesis-config", optional = true }
```
remember to switch to main branch before sending PR to upstream repo

scarmuega commented on mithril-common/src/chain_observer/pallas_observer.rs:
```diff
@@ -239,6 +231,41 @@ impl PallasChainObserver {
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainObserver failed to get genesis config")?;
+
+        let slots_per_kes_period = genesis_config[0].slots_per_kes_period as u64;
```
even if we're confident that the genesis_config array will have a `0` index I would not access the index directly since it could panic. It's better to use something like `get()` and return an Err in case the index it's not available.

falcucci committed:
chore: update dependency sources to `txpipe` repository

- update the `pallas-addresses`, `pallas-codec`, `pallas-network`, `pallas-primitives`, `pallas-traverse`, and `pallas-crypto` dependencies to use the `txpipe` repository and `main` branch instead of the `falcucci` repository and `feat/genesis-config` branch in `cargo.toml`

falcucci committed:
merge 'main'

falcucci committed:
refactor: improved error handling and configuration handling

- refactor the code to retrieve `slots_per_kes_period` from `config` instead of directly from `genesis_config`
- improve error handling by adding context to the error message
- adjust the calculation of `current_kes_period` to use the updated `slots_per_kes_period` calculation from `config`

falcucci committed:
chore: improve error messages for config extraction process

- change error message from obtaining genesis config to extracting the config.

falcucci committed:
fix: improve slots_per_kes_period validation and error messaging

- add validation for slots_per_kes_period being greater than 0
- update error message for slots_per_kes_period being 0

falcucci committed:
feat: refactor kes period calculation in chain module

- add a new async function `calculate_kes_period` that takes `&self, chain_point: point, slots_per_kes_period: u64` as arguments
- rename the function `calculate_kes_period` to `get_kes_period`
- modify code in the test mod to match the changes in `calculate_kes_period` function

falcucci committed:
chore: update specific points in `getchainpoint` request

- update the specific point value from 53536042 to 52851885 in the `getchainpoint` request.

falcucci committed:
chore: refactor method signature for `calculate_kes_period`

- add a comment for the `calculate_kes_period` method
- include `chain_point` and `slots_per_kes_period` as arguments at the method definition

falcucci committed:
feat: refactor return types and error handling in functions

- change return type of function from `u32` to `kesperiod`
- fix error handling and conversion in the function
- update test case with new chain point slot value

falcucci committed:
refactor: implementation logic for `slots_per_kes_period` being `0`

- refactor the `pallaschainobserver` implementation to handle `slots_per_kes_period` being `0` differently
- update the calculation of `current_kes_period` to handle the new logic for `slots_per_kes_period` being `0`

falcucci committed:
docs: enhance documentation for kes period calculation formula

- add detailed documentation for the `calculate_kes_period` function and its usage
- improve the explanation of the calculation formula for the kes period
- include an example with input values for the calculation formula

falcucci committed:
docs: correct variable descriptions

- fix a typo in the variable description from `s symbolizes` to `s represents`

falcucci committed:
refactor: standardize variable naming conventions across files

- update the comment and variable names in `mithril-common/src/chain_observer/pallas_observer.rs` to improve clarity.

falcucci committed:
Merge branch 'main' into feat/kes_period

falcucci committed:
refactor: chainobserver imports in infrastructure codebase

- remove unused imports related to chainobserver in infrastructure.rs
- update the import to use pallaschainobserver directly

falcucci committed:
merge 'main'

falcucci committed:
chore: update version number and dependencies

- update the version number in `mithril-common/cargo.toml` from `0.3.15` to `0.3.16`

falcucci committed:
docs: standardize code comments and examples clarity

- change code block syntax in comments to `no_run`
- update comments in code for example clarity

falcucci committed:
docs: revise comments for kes period calculations across files

- update comments for the `calculate current kes period` function in `mithril-common/src/chain_observer/pallas_observer.rs`

falcucci committed:
docs: update documentation and refactor code in observer module

- updated the documentation for the `calculate current kes period` function in `pallas_observer.rs`

falcucci committed:
feat: refactor `calculate_kes_period` function in pallas observer

- update `calculate_kes_period` function in `pallas_observer.rs` to include a new example with chain_point and slots_per_kes_period values
- remove the old example from the `calculate_kes_period` function in `pallas_observer.rs`
